[Name lookup] Use decl-based name lookup more regularly

DougGregor opened the pull request:
Avoid the use of the type checker in much of name lookup, relying on Decl-level information (not Type-level information) instead. There are several changes here that support each other to get back to a usable state:

* Use decl-based name lookup for unqualified lookup: replace uses of the `Type` form of `lookupQualified` (where the type was usually provided by `getSelfTypeInContext()`) with the "decl" form of the method, so we don't need to resolve types to perform unqualified name lookup.
* Use declaration-based name lookup in the conformance lookup table: similar to the above, allows us to determine the set of protocols to which a nominal type conforms using only Decl-level information.

With these changes, eliminate `LazyResolver::resolveExtensionForConformanceConstruction()` and some of the type-checker logic supporting it.

DougGregor committed:
[Name lookup] Use decl-based name lookup for unqualified lookup.

Within unqualified name lookup, replace uses of Type-based lookup
(formed from calls to getSelfTypeInContext()) with declaration-based
name lookup, so that name lookup doesn't depend directly on the type
checker.

The main oddity here is that we need to consider "Self: Foo"
requirements within a protocol extension during name lookup, so that
we will also look into "Foo". This is handled via the basic
name-lookup facilities that allow us to resolve TypeReprs to
declarations without going through the type checker.

slavapestov commented on lib/AST/NameLookup.cpp:
```diff
@@ -2201,8 +2277,6 @@ directReferencesForTypeRepr(Evaluator &evaluator,
-/// Retrieve the set of type declarations that are directly referenced from
-/// the given type.
 static DirectlyReferencedTypeDecls
 directReferencesForType(Type type) {
```
This looks like it could fit on one line

DougGregor replied:
Ok. Will do in a follow-up

DougGregor replied:
It ended up in https://github.com/apple/swift/pull/18562/files

DougGregor committed:
[AST] Use declaration-based name lookup in the conformance lookup table.

This avoids a call into the lazy resolver simply to figure out which
protocols a type conforms to.

DougGregor committed:
Remove LazyResolver::resolveExtensionForConformanceConstruction().

This is no longer used. NFC

DougGregor committed:
[Type checker] Eliminate TypeResolutionFlags::ResolveStructure.

It’s no longer introduced anywhere. NFC

DougGregor committed:
[AST] Eliminate the unused by ComputingInheritedProtocols in ProtocolDecl.

DougGregor committed:
[Conformance lookup table] Delete unnecessary call to bindExtension().

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor committed:
Fix test case for inconsequential output change

DougGregor committed:
[Name lookup] Turn the "Self bounds of a where clause" query into a request.

Name lookup within a protocol extension also looks into protocols and
superclasses on the right-hand side of Self constraints in the where
clause, e.g., "Self: Foo". Turn that function into a request to avoid
infinite recursion on invalid code.

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please test compiler performance

DougGregor commented:
@swift-ci please test compiler performance

DougGregor committed:
[AST Verifier] Remove verifyProtocolList().

The property that we are checking for here isn’t really local to the
nominal type or extension declaration, and triggers semantic queries
that aren’t needed elsewhere. Remove it, because it’s not adding value.

DougGregor commented:
@swift-ci please test compiler performance

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test compiler performance

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
[Type checker diags] Validate constructors when they don’t have interface types.

Fixes a diagnostics regression due to the name-shadowing changes.

DougGregor committed:
[Type checker] Don’t form a cycle when printing a request on an extension.

DougGregor committed:
[Conformance lookup table] Drop unnecessary LazyResolver parameters.

There is no point in threading LazyResolver parameters through this
data structure; we can recover the resolver in the one place it is 
needed.

DougGregor committed:
Eliminate two trivial uses of lookupQualified(Type, ...).

DougGregor committed:
[SourceKit] Test case fix-up because we’re not validating quite as eagerly.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility
